frame/support/src/storage/bounded_vec.rs

```diff
@@ -528,97 +535,155 @@ where
 
-impl<T, S> StorageDecodeLength for BoundedVec<T, S> {}
+impl<T, B, S: Get<B>> StorageDecodeLength for BoundedVec<T, B, S> {}
 
-impl<T, S: Get<u32>> StorageTryAppend<T> for BoundedVec<T, S> {
+impl<T, B: AtMost32BitUnsigned, S: Get<B>> StorageTryAppend<T> for BoundedVec<T, B, S> {
 	fn bound() -> usize {
-		S::get() as usize
+		let bound: usize = S::get().into().try_into().unwrap();
+		bound
 	}
 }
 
-impl<T, S> MaxEncodedLen for BoundedVec<T, S>
+impl<T, B, S> MaxEncodedLen for BoundedVec<T, B, S>
 where
 	T: MaxEncodedLen,
-	S: Get<u32>,
-	BoundedVec<T, S>: Encode,
+	B: AtMost32BitUnsigned,
+	S: Get<B>,
+	BoundedVec<T, B, S>: Encode,
 {
 	fn max_encoded_len() -> usize {
-		// BoundedVec<T, S> encodes like Vec<T> which encodes like [T], which is a compact u32
+		// BoundedVec<T, B, S> encodes like Vec<T> which encodes like [T], which is a compact u32
 		// plus each item in the slice:
 		// https://docs.substrate.io/v3/advanced/scale-codec
-		codec::Compact(S::get())
+		codec::Compact(S::get().into())
 			.encoded_size()
 			.saturating_add(Self::bound().saturating_mul(T::max_encoded_len()))
 	}
 }
 
-impl<I, T, Bound> TryCollect<BoundedVec<T, Bound>> for I
+impl<I, T, B, Bound> TryCollect<BoundedVec<T, B, Bound>> for I
 where
 	I: ExactSizeIterator + Iterator<Item = T>,
-	Bound: Get<u32>,
+	B: AtMost32BitUnsigned,
+	Bound: Get<B>,
 {
 	type Error = &'static str;
 
-	fn try_collect(self) -> Result<BoundedVec<T, Bound>, Self::Error> {
-		if self.len() > Bound::get() as usize {
+	fn try_collect(self) -> Result<BoundedVec<T, B, Bound>, Self::Error> {
+		if Bound::get().into().try_into().map_or(false, |v: usize| self.len() > v) {
 			Err("iterator length too big")
 		} else {
-			Ok(BoundedVec::<T, Bound>::unchecked_from(self.collect::<Vec<T>>()))
+			Ok(BoundedVec::<T, B, Bound>::unchecked_from(self.collect::<Vec<T>>()))
 		}
 	}
 }
 
 #[cfg(test)]
 pub mod test {
 	use super::*;
-	use crate::{bounded_vec, traits::ConstU32, Twox128};
+	use crate::{bounded_vec, traits::ConstU8, traits::ConstU16, traits::ConstU32, Twox128};
 	use sp_io::TestExternalities;
 
-	crate::generate_storage_alias! { Prefix, Foo => Value<BoundedVec<u32, ConstU32<7>>> }
-	crate::generate_storage_alias! { Prefix, FooMap => Map<(Twox128, u32), BoundedVec<u32, ConstU32<7>>> }
+	crate::generate_storage_alias! { Prefix, Foo => Value<BoundedVec<u32, u32, ConstU32<7>>> }
+	crate::generate_storage_alias! { Prefix, FooMap => Map<(Twox128, u32), BoundedVec<u32, u32, ConstU32<7>>> }
 	crate::generate_storage_alias! {
 		Prefix,
-		FooDoubleMap => DoubleMap<(Twox128, u32), (Twox128, u32), BoundedVec<u32, ConstU32<7>>>
+		FooDoubleMap => DoubleMap<(Twox128, u32), (Twox128, u32), BoundedVec<u32, u32, ConstU32<7>>>
 	}
 
 	#[test]
 	fn slide_works() {
-		let mut b: BoundedVec<u32, ConstU32<6>> = bounded_vec![0, 1, 2, 3, 4, 5];
+		let mut b: BoundedVec<u32, u32, ConstU32<6>> = bounded_vec![0, 1, 2, 3, 4, 5];
```

Comment:
what is the rationale for making _all_ tests work with all 3 possible bounds now? I would just add one tests that ensures different bound types can be used, and leave all the rest unchanged. 

But you can elaborate more on what scenario you are testing for, perhaps I am missing something.